### PR TITLE
Redirect to language version last used

### PR DESCRIPTION
**Pull request by Niharika29**

### CODE REVIEW

**Comment by MusikAnimal** [comment]
:+1: Looks good to me! Tested it on my local

File: src/App.php [hunk 1]
```diff
@@ -265,13 +265,21 @@ protected function configureRoutes( Slim $slim ) {
...
+					// See if we have a cookie indicating last version used
+					if ( isset( $_COOKIE['copypatrolLang'] ) ) {
+						$slim->redirectTo( 'home', [ 'wikiLang' => $_COOKIE['copypatrolLang'] ] );
+					} else {
+						// If no cookie, redirect to English version
+						$slim->redirectTo( 'home', [ 'wikiLang' => 'en' ] );
+					}
 					// Redirect root-URL requests to English.
 					$currentLang = $slim->i18nContext->getCurrentLanguage();
 					$slim->redirectTo( 'home', [ 'wikiLang' => $currentLang ] );
```
Comment (ricordisamoa) [inline src/App.php]: https://github.com/slimphp/Slim/blob/20a02782f76830b67ae56a5c08eb1f563c351a37/Slim/Slim.php#L1093
> This method immediately redirects to a new URL.

Comment (MusikAnimal) [inline src/App.php]: Is that not what we want?

Comment (ricordisamoa) [inline src/App.php]: I've not tested the code but it seems that the last `redirectTo` will never be called

Comment (MusikAnimal) [inline src/App.php]: Yep, I'm now realizing that as well. So I believe getting rid of the `else` block on lines 271-274 will do what we want. What do you think @Niharika29 ?

File: src/App.php [hunk 2]
```diff
@@ -265,13 +265,21 @@ protected function configureRoutes( Slim $slim ) {
 				$middleware['inject-user'], $middleware['set-environment'],
 			function () use ( $slim, $middleware, $routeConditions ) {
 				$slim->get( '', function() use ( $slim ) {
+					// See if we have a cookie indicating last version used
+					if ( isset( $_COOKIE['copypatrolLang'] ) ) {
```
Comment (samwilson) [inline src/App.php]: It doesn't make much difference here, but there's `$slim->getCookie()` that will do this as well. It returns null if there's no cookie.

File: src/App.php [hunk 3]
```diff
@@ -265,13 +265,21 @@ protected function configureRoutes( Slim $slim ) {
...
+						$slim->redirectTo( 'home', [ 'wikiLang' => 'en' ] );
+					}
 					// Redirect root-URL requests to English.
 					$currentLang = $slim->i18nContext->getCurrentLanguage();
 					$slim->redirectTo( 'home', [ 'wikiLang' => $currentLang ] );
 				} )->name( 'root' );
 				$slim->get( ':wikiLang',
 					function ( $wikiLang ) use ( $slim ) {
 						$page = new CopyPatrol( $slim );
+						setcookie( 'copypatrolLang', $wikiLang );
```
Comment (samwilson) [inline src/App.php]: Similarly to the above: `$slim->setCookie()` could also be used, I think.

File: src/App.php (hunk 1 above)
Comment (Niharika29) [inline src/App.php]: Yeah, my mistake. Fixing that now.

File: src/App.php (hunk 2 above)
Comment (Niharika29) [inline src/App.php]: I looked into that earlier. Slim only has session cookies whereas I thought we wanted ours to persist longer. I realize now I forgot to give it an expiry date. I'm thinking 30 days should be good.

File: src/App.php [hunk 4]
```diff
@@ -265,13 +265,18 @@ protected function configureRoutes( Slim $slim ) {
 				$middleware['inject-user'], $middleware['set-environment'],
 			function () use ( $slim, $middleware, $routeConditions ) {
 				$slim->get( '', function() use ( $slim ) {
-					// Redirect root-URL requests to English.
-					$currentLang = $slim->i18nContext->getCurrentLanguage();
```
Comment (Niharika29) [inline src/App.php]: Makes sense.

File: src/App.php [hunk 5]
```diff
@@ -265,13 +265,23 @@ protected function configureRoutes( Slim $slim ) {
...
+						$slim->redirectTo( 'home', [ 'wikiLang' => $_COOKIE['copypatrolLang'] ] );
+					} else {
+						// If no cookie, check if we support i18nContext's default language
+						$lang = $slim->i18nContext->getCurrentLanguage();
+						if ( in_array( $lang, [ 'en', 'fr' ] ) ) {
+							$slim->redirectTo( 'home', [ 'wikiLang' => $lang ] );
+						}
+						// We don't support i18nContext's current language, so redirect to en version
+						$slim->redirectTo( 'home', [ 'wikiLang' => 'en' ] );
+					}
```
Comment (Niharika29) [inline src/App.php]: It does look less understandable. I'd prefer sticking with the current version. :)

File: src/App.php [hunk 6]
```diff
@@ -265,13 +265,23 @@ protected function configureRoutes( Slim $slim ) {
...
-					// Redirect root-URL requests to English.
-					$currentLang = $slim->i18nContext->getCurrentLanguage();
-					$slim->redirectTo( 'home', [ 'wikiLang' => $currentLang ] );
+					// See if we have a cookie indicating last version used
+					if ( isset( $_COOKIE['copypatrolLang'] ) ) {
+						$slim->redirectTo( 'home', [ 'wikiLang' => $_COOKIE['copypatrolLang'] ] );
+					} else {
+						// If no cookie, check if we support i18nContext's default language
+						$lang = $slim->i18nContext->getCurrentLanguage();
+						if ( in_array( $lang, [ 'en', 'fr' ] ) ) {
```
Comment (kaldari) [inline src/App.php]: Can we use a config var for this? i.e. $supportedLanguages? Or even just a member var at the top of the file would work. Something so it isn't as easy to miss when we expand the languages further.

File: src/App.php [hunk 7]
```diff
@@ -49,6 +49,18 @@
 class App extends AbstractApp {
 
 	/**
+	 * @var array Languages supported by the app
+	 */
+	var $supportedLanguages = [ 'en', 'fr' ];
```
Comment (ricordisamoa) [inline src/App.php]:
```
 54 | ERROR | The var keyword must not be used to declare a property
 54 | ERROR | Visibility must be declared on property
    |       | "$supportedLanguages"
```

Comment (Niharika29) [inline src/App.php]: Thanks, I missed it. Fixed now.

File: src/App.php [hunk 8]
```diff
@@ -49,6 +49,18 @@
 class App extends AbstractApp {
 
 	/**
+	 * @var array Languages supported by the app
+	 */
+	public $supportedLanguages = [ 'en', 'fr' ];
```
Comment (ricordisamoa) [inline src/App.php]: May I ask what is the purpose of having a public property and a getter at the same time. Thank you.

**Pull request merged**